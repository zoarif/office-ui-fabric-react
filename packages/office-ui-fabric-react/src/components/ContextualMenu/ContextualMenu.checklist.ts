import { ChecklistStatus } from '../../demo/ComponentStatus/ComponentStatus.types';

export const ContextualMenuStatus = {
  keyboardAccessibilitySupport: ChecklistStatus.unknown,
  markupSupport: ChecklistStatus.unknown,
  highContrastSupport: ChecklistStatus.fail,
  rtlSupport: ChecklistStatus.fail,
  testCoverage: ChecklistStatus.good
};