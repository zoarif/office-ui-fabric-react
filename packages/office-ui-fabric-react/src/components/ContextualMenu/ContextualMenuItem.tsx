import * as React from 'react';
import { hasSubmenu, getIsChecked } from '../../utilities/contextualMenu/index';
import { getRTL } from '../../Utilities';
import { Icon } from '../../Icon';
import { IContextualMenuItemProps } from './ContextualMenuItem.types';

const renderItemIcon = (props: IContextualMenuItemProps) => {
  const {
    item,
    hasIcons,
    classNames
  } = props;

  // Only present to allow continued use of item.icon which is deprecated.
  const { iconProps, icon } = item;

  if (!hasIcons) {
    return null;
  }

  if (item.onRenderIcon) {
    return (
      item.onRenderIcon(props)
    );
  }

  if (iconProps) {
    return <Icon { ...iconProps } className={ classNames.icon } />;
  }

  return <Icon iconName={ icon } className={ classNames.icon } />;
};

const renderCheckMarkIcon = ({ onCheckmarkClick, item, classNames }: IContextualMenuItemProps) => {
  const isItemChecked = getIsChecked(item);
  if (onCheckmarkClick) {
    const onClick = (e: React.MouseEvent<HTMLElement>) => onCheckmarkClick(item, e);

    return (
      <Icon
        iconName={ isItemChecked ? 'CheckMark' : '' }
        className={ classNames.checkmarkIcon }
        onClick={ onClick }
      />
    );
  }
  return null;
};

const renderItemName = ({ item, classNames }: IContextualMenuItemProps) => {
  if (item.name) {
    return <span className={ classNames.label }>{ item.name }</span>;
  }
  return null;
};

const renderSubMenuIcon = ({ item, classNames }: IContextualMenuItemProps) => {
  if (hasSubmenu(item)) {
    return (
      <Icon
        iconName={ getRTL() ? 'ChevronLeft' : 'ChevronRight' }
        { ...item.submenuIconProps }
        className={ classNames.subMenuIcon }
      />
    );
  }
  return null;
};

export const ContextualMenuItem: React.StatelessComponent<IContextualMenuItemProps> = (props) => {
  const { item, classNames } = props;

  return (
    <div
      className={
        item.split ? classNames.linkContentMenu : classNames.linkContent
      }
    >
      { renderCheckMarkIcon(props) }
      { renderItemIcon(props) }
      { renderItemName(props) }
      { renderSubMenuIcon(props) }
    </div>
  );
};
