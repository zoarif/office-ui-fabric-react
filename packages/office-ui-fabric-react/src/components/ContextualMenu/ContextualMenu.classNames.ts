import { memoizeFunction } from '../../Utilities';
import { ITheme, mergeStyleSets } from '../../Styling';
import { getStyles as getContextualMenuStyles, getMenuItemStyles } from './ContextualMenu.styles';
import { IVerticalDividerClassNames } from '../Divider/VerticalDivider.types';
import { getDividerClassNames } from '../Divider/VerticalDivider.classNames';

export interface IContextualMenuClassNames {
  container: string;
  root: string;
  list: string;
  header: string;
  title: string;
}

export interface IMenuItemClassNames {
  item: string;
  divider: string;
  root: string;
  linkContent: string;
  icon: string;
  checkmarkIcon: string;
  subMenuIcon: string;
  label: string;
  splitContainer: string;
  splitPrimary: string;
  splitMenu: string;
  linkContentMenu: string;
}

export const getSplitButtonVerticalDividerClassNames = memoizeFunction((theme: ITheme): IVerticalDividerClassNames => {
  return mergeStyleSets(getDividerClassNames(theme), {
    divider: {
      height: 16,
      width: 1,
    }
  });
});

export const getContextualMenuClassNames = memoizeFunction((
  theme: ITheme,
  className?: string
): IContextualMenuClassNames => {

  const styles = getContextualMenuStyles(theme);

  return mergeStyleSets({
    container: [
      'ms-ContextualMenu-container',
      styles.container,
      className,
      [{
        selectors: {
          ':focus': { outline: 0 }
        }
      }]
    ],
    root: [
      'ms-ContextualMenu is-open',
      styles.root
    ],
    list: [
      'ms-ContextualMenu-list is-open',
      styles.list
    ],
    header: [
      'ms-ContextualMenu-header',
      styles.header
    ],
    title: styles.title
  });
});

export const getItemClassNames = memoizeFunction((
  theme: ITheme,
  disabled: boolean,
  expanded: boolean,
  checked: boolean,
  isAnchorLink: boolean,
  knownIcon: boolean,
  itemClassName?: string,
  dividerClassName?: string,
  iconClassName?: string,
  subMenuClassName?: string,
): IMenuItemClassNames => {

  const styles = getMenuItemStyles(theme);

  return mergeStyleSets({
    item: [
      'ms-ContextualMenu-item',
      styles.item,
      itemClassName,
    ],
    divider: [
      'ms-ContextualMenu-divider',
      styles.divider,
      dividerClassName,
    ],
    root: [
      'ms-ContextualMenu-link',
      styles.root,
      checked && [
        'is-checked',
        styles.rootChecked
      ],
      isAnchorLink && styles.anchorLink,
      expanded && [
        'is-expanded',
        styles.rootExpanded
      ],
      disabled && [
        'is-disabled',
        styles.rootDisabled
      ],
      !disabled && !expanded && [{
        selectors: {
          ':hover': styles.rootHovered,
          ':active': styles.rootPressed,
          '.ms-Fabric--isFocusVisible &:focus, .ms-Fabric--isFocusVisible &:focus:hover': styles.rootFocused,
          '.ms-Fabric--isFocusVisible &:hover': { background: 'inherit;' }
        }
      }],
    ],
    splitPrimary: [
      styles.root,
      checked && [
        'is-checked',
        styles.rootChecked
      ],
      disabled && [
        'is-disabled',
        styles.rootDisabled
      ],
      !disabled && !checked && [{
        selectors: {
          ':hover': styles.rootHovered,
          ':active': styles.rootPressed,
          '.ms-Fabric--isFocusVisible &:focus, .ms-Fabric--isFocusVisible &:focus:hover': styles.rootFocused,
          '.ms-Fabric--isFocusVisible &:hover': { background: 'inherit;' }
        }
      }]
    ],
    splitMenu: [
      styles.root,
      {
        width: 32
      },
      expanded && [
        'is-expanded',
        styles.rootExpanded
      ],
      disabled && [
        'is-disabled',
        styles.rootDisabled
      ],
      !disabled && !expanded && [{
        selectors: {
          ':hover': styles.rootHovered,
          ':active': styles.rootPressed,
          '.ms-Fabric--isFocusVisible &:focus, .ms-Fabric--isFocusVisible &:focus:hover': styles.rootFocused,
          '.ms-Fabric--isFocusVisible &:hover': { background: 'inherit;' }
        }
      }]
    ],
    linkContent: [
      'ms-ContextualMenu-linkContent',
      styles.linkContent
    ],
    linkContentMenu: [
      'ms-ContextualMenu-linkContent',
      styles.linkContent,
      {
        justifyContent: 'center',
      }
    ],
    icon: [
      'ms-ContextualMenu-icon',
      knownIcon && 'ms-ContextualMenu-iconColor ' && styles.iconColor,
      styles.icon,
      iconClassName,
      disabled && [
        'is-disabled',
        styles.iconDisabled
      ]
    ],
    checkmarkIcon: [
      'ms-ContextualMenu-checkmarkIcon',
      knownIcon && 'ms-ContextualMenu-checkmarkIcon ' && styles.checkmarkIcon,
      styles.icon,
      iconClassName,
    ],
    subMenuIcon: [
      'ms-ContextualMenu-submenuIcon',
      styles.subMenuIcon,
      subMenuClassName,
    ],
    label: [
      'ms-ContextualMenu-itemText',
      styles.label
    ],
    splitContainer: [
      styles.splitButtonFlexContainer,
      !disabled && !checked && [{
        selectors: {
          '.ms-Fabric--isFocusVisible &:focus, .ms-Fabric--isFocusVisible &:focus:hover': styles.rootFocused,
        }
      }]
    ]
  });
});