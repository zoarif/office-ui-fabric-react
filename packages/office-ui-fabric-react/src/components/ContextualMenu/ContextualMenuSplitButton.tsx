import * as React from 'react';
import {
  BaseComponent,
  assign,
  buttonProperties,
  getNativeProps,
  KeyCodes
} from '../../Utilities';
import { IContextualMenuItem } from '../../ContextualMenu';
import {
  IMenuItemClassNames,
  getSplitButtonVerticalDividerClassNames
} from './ContextualMenu.classNames';
import { ContextualMenuItem } from './ContextualMenuItem';

import { getIsChecked, isItemDisabled } from '../../utilities/contextualMenu/index';
import { VerticalDivider } from '../../Divider';
import { IContextualMenuSplitButtonProps } from './ContextualMenuSplitButton.types';

export interface IContextualMenuSplitButtonState { }

export class ContextualMenuSplitButton extends BaseComponent<IContextualMenuSplitButtonProps, IContextualMenuSplitButtonState> {
  private _splitButton: HTMLDivElement;

  public render(): JSX.Element | null {
    const {
      item,
      classNames,
      index,
      focusableElementIndex,
      totalItemCount,
      hasCheckmarks,
      hasIcons,
      onItemMouseLeave
    } = this.props;

    return (
      <div
        ref={ (splitButton: HTMLDivElement) => this._splitButton = splitButton }
        role={ 'button' }
        aria-labelledby={ item.ariaLabel }
        className={ classNames.splitContainer }
        aria-disabled={ isItemDisabled(item) }
        aria-haspopup={ true }
        aria-describedby={ item.ariaDescription }
        aria-checked={ item.isChecked || item.checked }
        aria-posinset={ focusableElementIndex + 1 }
        aria-setsize={ totalItemCount }
        onMouseEnter={ this._onItemMouseEnterPrimary }
        onMouseLeave={ onItemMouseLeave ? onItemMouseLeave.bind(this, { ...item, subMenuProps: null, items: null }) : undefined }
        onMouseMove={ this._onItemMouseMovePrimary }
        onKeyDown={ this._onItemKeyDown }
        onClick={ this._executeItemClick }
        tabIndex={ 0 }
        data-is-focusable={ true }
      >
        { this._renderSplitPrimaryButton(item, classNames, index, hasCheckmarks!, hasIcons!) }
        { this._renderSplitDivider(item) }
        { this._renderSplitIconButton(item, classNames, index) }
      </div >
    );
  }

  private _renderSplitPrimaryButton(item: IContextualMenuItem, classNames: IMenuItemClassNames, index: number, hasCheckmarks: boolean, hasIcons: boolean) {
    const isChecked: boolean | null | undefined = getIsChecked(item);
    const canCheck: boolean = isChecked !== null;
    const defaultRole = canCheck ? 'menuitemcheckbox' : 'menuitem';
    const {
      contextualMenuItemAs: ChildrenRenderer = ContextualMenuItem,
      onItemClick
    } = this.props;

    const itemProps = {
      key: item.key,
      disabled: isItemDisabled(item) || item.primaryDisabled,
      name: item.name,
      className: classNames.splitPrimary,
      role: item.role || defaultRole,
      canCheck: item.canCheck,
      isChecked: item.isChecked,
      checked: item.checked,
      icon: item.icon,
      iconProps: item.iconProps,
      'data-is-focusable': false,
      'aria-hidden': true
    } as IContextualMenuItem;
    return (
      <button {...getNativeProps(itemProps, buttonProperties) }>
        <ChildrenRenderer
          data-is-focusable={ false }
          item={ itemProps }
          classNames={ classNames }
          index={ index }
          onCheckmarkClick={ hasCheckmarks && onItemClick ? onItemClick.bind(this, item) : undefined }
          hasIcons={ hasIcons }
        />
      </button>
    );
  }

  private _renderSplitDivider(item: IContextualMenuItem) {
    const getDividerClassNames = item.getSplitButtonVerticalDividerClassNames || getSplitButtonVerticalDividerClassNames;
    return <VerticalDivider getClassNames={ getDividerClassNames } />;
  }

  private _renderSplitIconButton(item: IContextualMenuItem, classNames: IMenuItemClassNames, index: number) {
    const {
      contextualMenuItemAs: ChildrenRenderer = ContextualMenuItem,
      onItemMouseLeave,
      onItemMouseDown
    } = this.props;

    const itemProps = {
      onClick: this._onIconItemClick,
      disabled: isItemDisabled(item),
      className: classNames.splitMenu,
      subMenuProps: item.subMenuProps,
      submenuIconProps: item.submenuIconProps,
      split: true,
    } as IContextualMenuItem;

    const buttonProps = assign({}, getNativeProps(itemProps, buttonProperties), {
      onMouseEnter: this._onItemMouseEnterIcon,
      onMouseLeave: onItemMouseLeave ? onItemMouseLeave.bind(this, item) : undefined,
      onMouseDown: (ev: any) => onItemMouseDown ? onItemMouseDown(item, ev) : undefined,
      onMouseMove: this._onItemMouseMoveIcon,
      'data-is-focusable': false,
      'aria-hidden': true
    });

    return (
      <button {...buttonProps} >
        <ChildrenRenderer item={ itemProps } classNames={ classNames } index={ index } hasIcons={ false } />
      </button >
    );
  }

  private _onItemMouseEnterPrimary = (ev: React.MouseEvent<HTMLElement>): void => {
    const {
      item,
      onItemMouseEnter
    } = this.props;
    if (onItemMouseEnter) {
      onItemMouseEnter({ ...item, subMenuProps: undefined, items: undefined }, ev, this._splitButton);
    }
  }

  private _onItemMouseEnterIcon = (ev: React.MouseEvent<HTMLElement>): void => {
    const { item, onItemMouseEnter } = this.props;
    if (onItemMouseEnter) {
      onItemMouseEnter(item, ev, this._splitButton);
    }
  }

  private _onItemMouseMovePrimary = (ev: React.MouseEvent<HTMLElement>): void => {
    const {
      item,
      onItemMouseMove
    } = this.props;
    if (onItemMouseMove) {
      onItemMouseMove({ ...item, subMenuProps: undefined, items: undefined }, ev, this._splitButton);
    }
  }

  private _onItemMouseMoveIcon = (ev: React.MouseEvent<HTMLElement>): void => {
    const {
      item,
      onItemMouseMove
    } = this.props;
    if (onItemMouseMove) {
      onItemMouseMove(item, ev, this._splitButton);
    }
  }

  private _onIconItemClick = (ev: React.MouseEvent<HTMLElement>): void => {
    const { item, onItemClickBase } = this.props;
    if (onItemClickBase) {
      onItemClickBase(item, ev, (this._splitButton ? this._splitButton : ev.currentTarget) as HTMLElement);
    }
  }

  private _executeItemClick = (ev: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLElement>): void => {
    const {
      item,
      executeItemClick
    } = this.props;

    if (item.disabled || item.isDisabled) {
      return;
    }

    if (executeItemClick) {
      executeItemClick(item, ev);
    }
  }

  private _onItemKeyDown = (ev: React.KeyboardEvent<HTMLElement>): void => {
    const { item, onItemKeyDown } = this.props;
    if (ev.which === KeyCodes.enter) {
      this._executeItemClick(ev);
      ev.preventDefault();
      ev.stopPropagation();
    } else if (onItemKeyDown) {
      onItemKeyDown(item, ev);
    }
  }
}
