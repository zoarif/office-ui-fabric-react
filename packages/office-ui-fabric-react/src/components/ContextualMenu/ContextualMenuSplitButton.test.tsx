import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { IContextualMenuItem } from './ContextualMenu.types';
import { IMenuItemClassNames } from './ContextualMenu.classNames';
import { ContextualMenuSplitButton } from './ContextualMenuSplitButton';
import { IContextualMenuSplitButtonProps } from './ContextualMenuSplitButton.types';

describe('ContextualMenuSplitButton', () => {
  describe('creates a normal split button', () => {
    let menuItem: IContextualMenuItem;
    let menuClassNames: IMenuItemClassNames;

    let wrapper: ShallowWrapper<IContextualMenuSplitButtonProps, {}>;
    beforeEach(() => {
      menuItem = { key: '123' };
      menuClassNames = getMenuItemClassNames();
      wrapper = shallow(
        <ContextualMenuSplitButton
          item={ menuItem }
          classNames={ menuClassNames }
          index={ 0 }
          focusableElementIndex={ 0 }
          totalItemCount={ 1 }
        />
      );
    });

    it('renders the contextual menu split button correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});

function getMenuItemClassNames(): IMenuItemClassNames {
  return {
    item: 'item',
    divider: '---',
    root: 'root',
    linkContent: 'linkContent',
    icon: 'icon',
    checkmarkIcon: 'checkmarkIcon',
    subMenuIcon: 'subMenuIcon',
    label: 'label',
    splitContainer: 'splitContainer',
    splitPrimary: 'splitPrimary',
    splitMenu: 'splitMenu',
    linkContentMenu: 'linkContentMenu',
  };
}
