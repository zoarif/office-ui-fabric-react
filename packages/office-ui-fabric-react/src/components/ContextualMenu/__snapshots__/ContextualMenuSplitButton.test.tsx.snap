// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextualMenuSplitButton creates a normal split button renders the contextual menu split button correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ContextualMenuSplitButton
    classNames={
      Object {
        "checkmarkIcon": "checkmarkIcon",
        "divider": "---",
        "icon": "icon",
        "item": "item",
        "label": "label",
        "linkContent": "linkContent",
        "linkContentMenu": "linkContentMenu",
        "root": "root",
        "splitContainer": "splitContainer",
        "splitMenu": "splitMenu",
        "splitPrimary": "splitPrimary",
        "subMenuIcon": "subMenuIcon",
      }
    }
    focusableElementIndex={0}
    index={0}
    item={
      Object {
        "key": "123",
      }
    }
    totalItemCount={1}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "aria-checked": undefined,
      "aria-describedby": undefined,
      "aria-disabled": false,
      "aria-haspopup": true,
      "aria-labelledby": undefined,
      "aria-posinset": 1,
      "aria-setsize": 1,
      "children": Array [
        <button
          aria-hidden={true}
          checked={undefined}
          className="splitPrimary"
          data-is-focusable={false}
          disabled={undefined}
          icon={undefined}
          name={undefined}
          role="menuitem"
        >
          <Unknown
            classNames={
              Object {
                "checkmarkIcon": "checkmarkIcon",
                "divider": "---",
                "icon": "icon",
                "item": "item",
                "label": "label",
                "linkContent": "linkContent",
                "linkContentMenu": "linkContentMenu",
                "root": "root",
                "splitContainer": "splitContainer",
                "splitMenu": "splitMenu",
                "splitPrimary": "splitPrimary",
                "subMenuIcon": "subMenuIcon",
              }
            }
            data-is-focusable={false}
            hasIcons={undefined}
            index={0}
            item={
              Object {
                "aria-hidden": true,
                "canCheck": undefined,
                "checked": undefined,
                "className": "splitPrimary",
                "data-is-focusable": false,
                "disabled": undefined,
                "icon": undefined,
                "iconProps": undefined,
                "isChecked": undefined,
                "key": "123",
                "name": undefined,
                "role": "menuitem",
              }
            }
            onCheckmarkClick={undefined}
          />
        </button>,
        <Unknown
          getClassNames={[Function]}
        />,
        <button
          aria-hidden={true}
          className="splitMenu"
          data-is-focusable={false}
          disabled={false}
          onClick={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={undefined}
          onMouseMove={[Function]}
        >
          <Unknown
            classNames={
              Object {
                "checkmarkIcon": "checkmarkIcon",
                "divider": "---",
                "icon": "icon",
                "item": "item",
                "label": "label",
                "linkContent": "linkContent",
                "linkContentMenu": "linkContentMenu",
                "root": "root",
                "splitContainer": "splitContainer",
                "splitMenu": "splitMenu",
                "splitPrimary": "splitPrimary",
                "subMenuIcon": "subMenuIcon",
              }
            }
            hasIcons={false}
            index={0}
            item={
              Object {
                "className": "splitMenu",
                "disabled": false,
                "onClick": [Function],
                "split": true,
                "subMenuProps": undefined,
                "submenuIconProps": undefined,
              }
            }
          />
        </button>,
      ],
      "className": "splitContainer",
      "data-is-focusable": true,
      "onClick": [Function],
      "onKeyDown": [Function],
      "onMouseEnter": [Function],
      "onMouseLeave": undefined,
      "onMouseMove": [Function],
      "role": "button",
      "tabIndex": 0,
    },
    "ref": [Function],
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "aria-hidden": true,
          "checked": undefined,
          "children": <Unknown
            classNames={
              Object {
                "checkmarkIcon": "checkmarkIcon",
                "divider": "---",
                "icon": "icon",
                "item": "item",
                "label": "label",
                "linkContent": "linkContent",
                "linkContentMenu": "linkContentMenu",
                "root": "root",
                "splitContainer": "splitContainer",
                "splitMenu": "splitMenu",
                "splitPrimary": "splitPrimary",
                "subMenuIcon": "subMenuIcon",
              }
            }
            data-is-focusable={false}
            hasIcons={undefined}
            index={0}
            item={
              Object {
                "aria-hidden": true,
                "canCheck": undefined,
                "checked": undefined,
                "className": "splitPrimary",
                "data-is-focusable": false,
                "disabled": undefined,
                "icon": undefined,
                "iconProps": undefined,
                "isChecked": undefined,
                "key": "123",
                "name": undefined,
                "role": "menuitem",
              }
            }
            onCheckmarkClick={undefined}
          />,
          "className": "splitPrimary",
          "data-is-focusable": false,
          "disabled": undefined,
          "icon": undefined,
          "name": undefined,
          "role": "menuitem",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "classNames": Object {
              "checkmarkIcon": "checkmarkIcon",
              "divider": "---",
              "icon": "icon",
              "item": "item",
              "label": "label",
              "linkContent": "linkContent",
              "linkContentMenu": "linkContentMenu",
              "root": "root",
              "splitContainer": "splitContainer",
              "splitMenu": "splitMenu",
              "splitPrimary": "splitPrimary",
              "subMenuIcon": "subMenuIcon",
            },
            "data-is-focusable": false,
            "hasIcons": undefined,
            "index": 0,
            "item": Object {
              "aria-hidden": true,
              "canCheck": undefined,
              "checked": undefined,
              "className": "splitPrimary",
              "data-is-focusable": false,
              "disabled": undefined,
              "icon": undefined,
              "iconProps": undefined,
              "isChecked": undefined,
              "key": "123",
              "name": undefined,
              "role": "menuitem",
            },
            "onCheckmarkClick": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "getClassNames": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "aria-hidden": true,
          "children": <Unknown
            classNames={
              Object {
                "checkmarkIcon": "checkmarkIcon",
                "divider": "---",
                "icon": "icon",
                "item": "item",
                "label": "label",
                "linkContent": "linkContent",
                "linkContentMenu": "linkContentMenu",
                "root": "root",
                "splitContainer": "splitContainer",
                "splitMenu": "splitMenu",
                "splitPrimary": "splitPrimary",
                "subMenuIcon": "subMenuIcon",
              }
            }
            hasIcons={false}
            index={0}
            item={
              Object {
                "className": "splitMenu",
                "disabled": false,
                "onClick": [Function],
                "split": true,
                "subMenuProps": undefined,
                "submenuIconProps": undefined,
              }
            }
          />,
          "className": "splitMenu",
          "data-is-focusable": false,
          "disabled": false,
          "onClick": [Function],
          "onMouseDown": [Function],
          "onMouseEnter": [Function],
          "onMouseLeave": undefined,
          "onMouseMove": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "classNames": Object {
              "checkmarkIcon": "checkmarkIcon",
              "divider": "---",
              "icon": "icon",
              "item": "item",
              "label": "label",
              "linkContent": "linkContent",
              "linkContentMenu": "linkContentMenu",
              "root": "root",
              "splitContainer": "splitContainer",
              "splitMenu": "splitMenu",
              "splitPrimary": "splitPrimary",
              "subMenuIcon": "subMenuIcon",
            },
            "hasIcons": false,
            "index": 0,
            "item": Object {
              "className": "splitMenu",
              "disabled": false,
              "onClick": [Function],
              "split": true,
              "subMenuProps": undefined,
              "submenuIconProps": undefined,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "aria-checked": undefined,
        "aria-describedby": undefined,
        "aria-disabled": false,
        "aria-haspopup": true,
        "aria-labelledby": undefined,
        "aria-posinset": 1,
        "aria-setsize": 1,
        "children": Array [
          <button
            aria-hidden={true}
            checked={undefined}
            className="splitPrimary"
            data-is-focusable={false}
            disabled={undefined}
            icon={undefined}
            name={undefined}
            role="menuitem"
          >
            <Unknown
              classNames={
                Object {
                  "checkmarkIcon": "checkmarkIcon",
                  "divider": "---",
                  "icon": "icon",
                  "item": "item",
                  "label": "label",
                  "linkContent": "linkContent",
                  "linkContentMenu": "linkContentMenu",
                  "root": "root",
                  "splitContainer": "splitContainer",
                  "splitMenu": "splitMenu",
                  "splitPrimary": "splitPrimary",
                  "subMenuIcon": "subMenuIcon",
                }
              }
              data-is-focusable={false}
              hasIcons={undefined}
              index={0}
              item={
                Object {
                  "aria-hidden": true,
                  "canCheck": undefined,
                  "checked": undefined,
                  "className": "splitPrimary",
                  "data-is-focusable": false,
                  "disabled": undefined,
                  "icon": undefined,
                  "iconProps": undefined,
                  "isChecked": undefined,
                  "key": "123",
                  "name": undefined,
                  "role": "menuitem",
                }
              }
              onCheckmarkClick={undefined}
            />
          </button>,
          <Unknown
            getClassNames={[Function]}
          />,
          <button
            aria-hidden={true}
            className="splitMenu"
            data-is-focusable={false}
            disabled={false}
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={undefined}
            onMouseMove={[Function]}
          >
            <Unknown
              classNames={
                Object {
                  "checkmarkIcon": "checkmarkIcon",
                  "divider": "---",
                  "icon": "icon",
                  "item": "item",
                  "label": "label",
                  "linkContent": "linkContent",
                  "linkContentMenu": "linkContentMenu",
                  "root": "root",
                  "splitContainer": "splitContainer",
                  "splitMenu": "splitMenu",
                  "splitPrimary": "splitPrimary",
                  "subMenuIcon": "subMenuIcon",
                }
              }
              hasIcons={false}
              index={0}
              item={
                Object {
                  "className": "splitMenu",
                  "disabled": false,
                  "onClick": [Function],
                  "split": true,
                  "subMenuProps": undefined,
                  "submenuIconProps": undefined,
                }
              }
            />
          </button>,
        ],
        "className": "splitContainer",
        "data-is-focusable": true,
        "onClick": [Function],
        "onKeyDown": [Function],
        "onMouseEnter": [Function],
        "onMouseLeave": undefined,
        "onMouseMove": [Function],
        "role": "button",
        "tabIndex": 0,
      },
      "ref": [Function],
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "aria-hidden": true,
            "checked": undefined,
            "children": <Unknown
              classNames={
                Object {
                  "checkmarkIcon": "checkmarkIcon",
                  "divider": "---",
                  "icon": "icon",
                  "item": "item",
                  "label": "label",
                  "linkContent": "linkContent",
                  "linkContentMenu": "linkContentMenu",
                  "root": "root",
                  "splitContainer": "splitContainer",
                  "splitMenu": "splitMenu",
                  "splitPrimary": "splitPrimary",
                  "subMenuIcon": "subMenuIcon",
                }
              }
              data-is-focusable={false}
              hasIcons={undefined}
              index={0}
              item={
                Object {
                  "aria-hidden": true,
                  "canCheck": undefined,
                  "checked": undefined,
                  "className": "splitPrimary",
                  "data-is-focusable": false,
                  "disabled": undefined,
                  "icon": undefined,
                  "iconProps": undefined,
                  "isChecked": undefined,
                  "key": "123",
                  "name": undefined,
                  "role": "menuitem",
                }
              }
              onCheckmarkClick={undefined}
            />,
            "className": "splitPrimary",
            "data-is-focusable": false,
            "disabled": undefined,
            "icon": undefined,
            "name": undefined,
            "role": "menuitem",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "classNames": Object {
                "checkmarkIcon": "checkmarkIcon",
                "divider": "---",
                "icon": "icon",
                "item": "item",
                "label": "label",
                "linkContent": "linkContent",
                "linkContentMenu": "linkContentMenu",
                "root": "root",
                "splitContainer": "splitContainer",
                "splitMenu": "splitMenu",
                "splitPrimary": "splitPrimary",
                "subMenuIcon": "subMenuIcon",
              },
              "data-is-focusable": false,
              "hasIcons": undefined,
              "index": 0,
              "item": Object {
                "aria-hidden": true,
                "canCheck": undefined,
                "checked": undefined,
                "className": "splitPrimary",
                "data-is-focusable": false,
                "disabled": undefined,
                "icon": undefined,
                "iconProps": undefined,
                "isChecked": undefined,
                "key": "123",
                "name": undefined,
                "role": "menuitem",
              },
              "onCheckmarkClick": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "getClassNames": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "aria-hidden": true,
            "children": <Unknown
              classNames={
                Object {
                  "checkmarkIcon": "checkmarkIcon",
                  "divider": "---",
                  "icon": "icon",
                  "item": "item",
                  "label": "label",
                  "linkContent": "linkContent",
                  "linkContentMenu": "linkContentMenu",
                  "root": "root",
                  "splitContainer": "splitContainer",
                  "splitMenu": "splitMenu",
                  "splitPrimary": "splitPrimary",
                  "subMenuIcon": "subMenuIcon",
                }
              }
              hasIcons={false}
              index={0}
              item={
                Object {
                  "className": "splitMenu",
                  "disabled": false,
                  "onClick": [Function],
                  "split": true,
                  "subMenuProps": undefined,
                  "submenuIconProps": undefined,
                }
              }
            />,
            "className": "splitMenu",
            "data-is-focusable": false,
            "disabled": false,
            "onClick": [Function],
            "onMouseDown": [Function],
            "onMouseEnter": [Function],
            "onMouseLeave": undefined,
            "onMouseMove": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "classNames": Object {
                "checkmarkIcon": "checkmarkIcon",
                "divider": "---",
                "icon": "icon",
                "item": "item",
                "label": "label",
                "linkContent": "linkContent",
                "linkContentMenu": "linkContentMenu",
                "root": "root",
                "splitContainer": "splitContainer",
                "splitMenu": "splitMenu",
                "splitPrimary": "splitPrimary",
                "subMenuIcon": "subMenuIcon",
              },
              "hasIcons": false,
              "index": 0,
              "item": Object {
                "className": "splitMenu",
                "disabled": false,
                "onClick": [Function],
                "split": true,
                "subMenuProps": undefined,
                "submenuIconProps": undefined,
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
